feature(toolbar) - addition of toolbar component

gavin-everett-genesys opened the pull request:
WIP

katie-bobbe-genesys commented on src/components/beta/gux-toolbar/i18n/en.json:
```diff
@@ -0,0 +1,3 @@
+{
+  "action": "{label}"
```
The label is provided by the consumer of the component and we cannot translate it for them. The `label` here is just a variable that cannot be translated.  There is no need to use i18n here, and you can just refer to `this.label` if you need to use the string in your component

katie-bobbe-genesys commented on src/components/beta/gux-toolbar/gux-toolbar-action/gux-toolbar-action.tsx:
```diff
@@ -0,0 +1,142 @@
+
+  render(): JSX.Element {
+    return (
+      <button
```
Instead of styling a button from scratch, you should just be able to use `gux-button` here. The styles in `gux-toolbar-action-primary` can be achieved by adding the attribute `accent="primary"` to `gux-button`. Also, you can remove the `tabindex` attribute as well

gavin-everett-genesys replied:
I think I confused myself when doing this I remember we were saying not to use gux-button as people could use the extra functionalities associated with it if it was in the lightDOM but seeing as it sits in the shadowDOM its ok 👍

katie-bobbe-genesys commented on src/components/beta/gux-toolbar/gux-toolbar-action/gux-toolbar-action.tsx:
```diff
@@ -0,0 +1,142 @@
+    }
+  }
+
+  private renderSrText(): JSX.Element {
```
There is no need for this block, because the `gux-icon` has a screen reader text property that will convey the label content. Adding this extra text just will cause the screen reader to read the same thing twice

katie-bobbe-genesys commented on src/components/beta/gux-toolbar/gux-toolbar-action/gux-toolbar-action.tsx:
```diff
@@ -0,0 +1,142 @@
+  icon: string;
+
+  @State()
+  sectionSpacing: boolean = false;
```
Something like `iconOnly` might be more descriptive of this state

katie-bobbe-genesys commented on src/components/beta/gux-toolbar/gux-toolbar-action/gux-toolbar-action.tsx:
```diff
@@ -0,0 +1,142 @@
+  private renderActionTitle(): JSX.Element {
+    if (!this.sectionSpacing) {
+      return (
+        <gux-tooltip-title>
```
Should the `gux-tooltip-title` contain both the text and the icon?

katie-bobbe-genesys commented on src/components/beta/gux-toolbar/example.html:
```diff
@@ -0,0 +1,28 @@
+  </div>
+
+  <div slot="permanent">
+    <gux-toolbar-action icon="delete">Delete</gux-toolbar-action>
```
Have we considered having default toolbar actions? The user could specify an `action-type` then we would provide the appropriate icon and text along with localizations. For example, maybe something like this:
`<gux-toolbar-action action-type="delete"></gux-toolbar-action>`
`<gux-toolbar-action action-type="filter"></gux-toolbar-action>`

And then we could also support custom action types, maybe something like this:
```
<gux-toolbar-action>
     <gux-icon icon-name="subtract" decorative></gux-icon>
     No Impact
 </gux-toolbar-action>
```

gavin-everett-genesys replied:
I like this idea, ye there is currently around six default actions outlined in the figma design so giving the user the option for a default or custom would definitely be good 👍

gavin-everett-genesys commented:
I think this is ready for a re-review now. I am still waiting on a few items from the design team which include,

- A max-width for the toolbar actions so I can implement the tooltip-title.
- A revert icon needs to be created currently there is none as far as I am aware.

gavin-everett-genesys commented:
Updates since last review : 

- Addition of Katies idea to have the ability of default actions bundled and also the ability to supply custom actions. A prop called `action` can be added to the `toolbar-action` and the end user can choose from 1 of 6 defaults. Otherwise they can slot a custom action.
- Utilising `gux-button` instead of a standard HTML button in `toolbar-action`
- Minor document changes.

MattCheely commented on src/components/beta/gux-toolbar/example.html:
```diff
@@ -0,0 +1,43 @@
+
+<h2>Toolbar Example</h2>
+
+<gux-toolbar-beta>
```
Should this be `gux-table-toolbar-beta`? It was my impression that this toolbar was meant specifically for tables.

MattCheely reviewed (commented):
Did you try implementing the part from the responsive design where the buttons collapse into the menu? 

I'm not sure if that's even feasible, just curious if you looked into it at all.

gavin-everett-genesys commented on src/components/beta/gux-toolbar/gux-toolbar.tsx:
```diff
@@ -0,0 +1,104 @@
+      const layoutContainerWidth = el.clientWidth;
+
+      if (layoutContainerWidth == MIN_SPACING) {
+        this.emitLayoutChange.emit(true);
```
From the collaboration call on Friday, maybe we should follow the same approach as the `pagination-beta` where we would still use some sort of functionality that will remove the icons when the max-width is 72px but also allow the end user to supply a prop with the layout they wish to use. This might help us get around the issue where we wont know how many actions the user is supplying and what languages they are using on the toolbar. The prop could be called `layout` and you can supply two values(I am not fully certain on the names) `advanced` and `iconOnly`.

MattCheely commented on src/components/beta/gux-toolbar/gux-toolbar.tsx:
```diff
@@ -0,0 +1,90 @@
+  root: HTMLElement;
+
+  @Event()
+  layout: EventEmitter<string>;
```
This should probably be `EventEmitter<GuxToolbarLayout>`

MattCheely commented on src/components/beta/gux-toolbar/gux-toolbar.tsx:
```diff
@@ -0,0 +1,90 @@
+  displayedLayout: GuxToolbarLayout;
+
+  @Watch('displayedLayout')
+  onLayoutChange(layoutType: string) {
```
I'd use `layoutType: GuxToolbarLayout`

MattCheely commented on src/components/beta/gux-toolbar/gux-toolbar.types.ts:
```diff
@@ -0,0 +1 @@
+export type GuxToolbarLayout = 'advanced' | 'iconOnly';
```
`advanced` doesn't seem like the best name in this case. Maybe `full`?

MattCheely commented on src/components/beta/gux-toolbar/gux-toolbar.tsx:
```diff
@@ -0,0 +1,90 @@
+
+@Component({
+  styleUrl: 'gux-toolbar.less',
+  tag: 'gux-toolbar-beta',
```
I think this should be `gux-table-toolbar-beta` unless UX indicates this is more general purpose than what it was originally designed for.

MattCheely commented on src/components/beta/gux-toolbar/gux-toolbar-constants.ts:
```diff
@@ -0,0 +1,9 @@
+/**
+ * If the spacing between the left and right controls is greater than 330 switch back to labels.
+ */
+export const ICON_ONLY_LAYOUT = 330;
```
I don't think there's a fixed size you can use for this strategy, as the number goes up the more buttons you have in the component. If you take the current example and just add 3 or 4 additional buttons and resize a bit, it's not hard to get the buttons to overflow the toolbar when growing the component.  I'm going to experiment with some ideas for another approach and I'll let you know if  I come up with something good.

rinckemd commented on src/components/beta/gux-toolbar/gux-toolbar-menu-button/gux-toolbar-menu-button-popover.less/readme.md:
```diff
@@ -0,0 +1,38 @@
+# gux-toolbar-menu-button-popover
```
Did you mean to have ".less" in the folder name?

rinckemd commented on src/components/beta/gux-toolbar/gux-toolbar.less:
```diff
@@ -0,0 +1,62 @@
+  width: 100%;
+  height: 48px;
+  padding: @spacing-xs;
+  background-color: @gux-grey-90;
```
Should this background color be grey-100? That's what I see in figma, but I not really in the loop.

rinckemd commented on src/components/beta/gux-toolbar/gux-toolbar-action/gux-toolbar-action.tsx:
```diff
@@ -0,0 +1,112 @@
+    const slotAssignedNodes = (
+      event.composedPath()[0] as HTMLSlotElement
+    ).assignedNodes();
+    this.label = slotAssignedNodes
```
I see a "[gux-icon] No screenreader-text provided. " error in console for markup like "<gux-toolbar-action icon="cog">Customize</gux-toolbar-action>". Seems like the label is set after the gux-icon is initially rendered?

katie-bobbe-genesys commented on src/components/beta/gux-toolbar/gux-toolbar-menu-button/gux-toolbar-menu-button.tsx:
```diff
@@ -0,0 +1,143 @@
+      <gux-toolbar-menu-button-popover expanded={this.isOpen}>
+        <div slot="target" class="gux-toolbar-menu-container">
+          <gux-button-slot-beta class="gux-dropdown-button">
+            <button
```
This button needs some kind of accessible text. I would remove the `decorative` property from `gux-icon` and add the `screenreader-text` property. The text should be something like "more actions" or "additional actions" I think.

katie-bobbe-genesys commented on src/components/beta/gux-toolbar/example.html:
```diff
@@ -0,0 +1,43 @@
+
+<gux-toolbar-beta>
+  <div slot="searchFilter">
+    <gux-form-field-search>
```
The form field search will throw an error if a label isn't added. Here is how to add visibly hidden label
```
    <gux-form-field-search label-position="screenreader">
      <input
        slot="input"
        class="search"
        type="search"
        name="a-3"
        placeholder="Enter search"
      />      
     <label slot="label">Toolbar Search</label>
    </gux-form-field-search>
```

katie-bobbe-genesys commented on src/components/beta/gux-toolbar/gux-toolbar-action/gux-toolbar-action.tsx:
```diff
@@ -0,0 +1,101 @@
+  root: HTMLElement;
+
+  @State()
+  label: string;
```
I noticed that there is an issue with icon only buttons not having screen reader accessible text in the `gux-toolbar-action` component. 

I think this component can be simplified if we don't try to compute the `label` from the slot text content. Instead of conditionally rendering the text content, we can visually hide the text when it is `icon-only`. If we do it this way, the visually hidden text content is still available to the screen reader and you don't need to add the `screenreader-text` to the icons. I have pasted some example code for this approach below. Reach out if you have any questions or concerns about this approach
gux-toolbar-action.tsx:
```
import { Component, Element, JSX, h, Prop } from '@stencil/core';
import { trackComponent } from '../../../../usage-tracking';
import { GuxToolbarActionTypes } from './gux-toolbar-action.types';


@Component({
  styleUrl: 'gux-toolbar-action.less',
  tag: 'gux-toolbar-action',
  shadow: true
})
export class GuxToolbarAction {

  /**
   * Reference to the host element.
   */
  @Element()
  root: HTMLElement;

  @Prop()
  primary: boolean;

  @Prop()
  icon: string;

  @Prop()
  action: GuxToolbarActionTypes;

  @Prop()
  iconOnly: boolean = false;

  private renderActionTitle(): JSX.Element {
    return (
      <span class={{ 'gux-sr-only': this.iconOnly }}>
        <slot />
      </span>
    ) as JSX.Element;
  }

  private renderIcon(): JSX.Element {
    if (this.icon) {
      return (
        <gux-icon iconName={this.icon} decorative />
      ) as JSX.Element;
    }
  }

  private renderActionType(): JSX.Element {
    return (
      <gux-button accent={this.primary ? 'primary' : 'secondary'} type="button">
        <gux-icon
          iconName={this.action == 'revert' ? 'reset' : this.action}
          decorative
        />
        <span class={{ 'gux-sr-only': this.iconOnly }}>
          {this.action.charAt(0).toUpperCase() + this.action.slice(1)}
        </span>
      </gux-button>
    ) as JSX.Element;
  }

  private renderCustomActionType(): JSX.Element {
    return (
      <gux-button accent={this.primary ? 'primary' : 'secondary'} type="button">
        {this.renderIcon()}
        {this.renderActionTitle()}
      </gux-button>
    ) as JSX.Element;
  }

  async componentWillLoad(): Promise<void> {
    trackComponent(this.root);
  }

  render(): JSX.Element {
    return this.action
      ? this.renderActionType()
      : this.renderCustomActionType();
  }
}
```
gux-toolbar-action.less (new file):
```
@import (reference) '../../../../style/typography.less';

.gux-sr-only {
  .gux-sr-only-clip();
}
```

gavin-everett-genesys commented:
> Accessibility is looking good now 👍 I did have a question around the width of the search input. I might expect that the width of the search input should be set by the component itself. I noticed that the width of the search input was being set in the example file. If I comment out this style, the search input resize behavior is strange:
> 
> Click to see gif example
> Should we expect users to style the search input width themselves? I personally think the component should probably set the min-width of the search input.

I will confirm this with UX and see what the expectation for it is. Id agree that there probably should be a min-width set.

gavin-everett-genesys commented:
The `table-toolbar` is ready for final review.

Items that have been added since, 

- Condensed menu. Actions will fold into menu button when view gets smaller.
- A min-width of 145px has been applied to the `form-field-search` with a flex-basis of 260px as requested by UX. Since the component has shadowDOM enabled I am wondering what is the best way to style the component. I currently have a method which just sets the style in the `componentWillLoad()`. I looked at the `::part` pseudo-element but I dont think that will work.
- Some e2e and spec tests.
- Rename of localisations.

katie-bobbe-genesys commented:
I noticed that when resizing the window to be wider, the condensed menu changes to the expanded menu, skipping the intermediate menu size. 
<details>
<summary>click to see gif</summary>

[![Image from Gyazo](https://i.gyazo.com/2e40a39fccb1882aaec6ddbc56820aec.gif)](https://gyazo.com/2e40a39fccb1882aaec6ddbc56820aec)
</details>

Also, the intermediate menu size only seems to show on resize

<details>
<summary>click to see gif</summary>

[![Image from Gyazo](https://i.gyazo.com/e005511c75486ee160c0769c7b743e41.gif)](https://gyazo.com/e005511c75486ee160c0769c7b743e41)
</details

daragh-king-genesys commented on src/components/beta/gux-table-toolbar/example.html:
```diff
@@ -0,0 +1,48 @@
+      <gux-list-item onclick="notify(event)">Extract</gux-list-item>
+    </gux-table-toolbar-menu-button>
+    <gux-table-toolbar-action onclick="notify(event)" primary icon="add"
+      >Add</gux-table-toolbar-action
```
Not a big fan of this api where there are some hardcoded actions and then a custom option that uses other props. It seems to me like at least two separate components are mixed together for no real reason.

MattCheely replied:
I think maybe it got changed, but the idea is that there are a set of pre-made actions that should always have the same text/icon pairing set out in the design system.  I think it's worth providing short cuts to those in some way so that devs don't have to remember to always get the pairing right, or have to manually update if the design system changes the icon that should be used.

daragh-king-genesys commented on src/components/beta/gux-table-toolbar/gux-table-toolbar-action/gux-table-toolbar-action.tsx:
```diff
@@ -0,0 +1,90 @@
+  root: HTMLElement;
+
+  @Prop()
+  primary: boolean;
```
I would use 
```
@Prop()
accent: 'primary' | 'secondary' = 'secondary';
```

daragh-king-genesys commented on src/components/beta/gux-table-toolbar/gux-table-toolbar-action/gux-table-toolbar-action.tsx:
```diff
@@ -0,0 +1,90 @@
+  primary: boolean;
+
+  @Prop()
+  icon: string;
```
why did you go with a prop here over a slot?

daragh-king-genesys commented on src/components/beta/gux-table-toolbar/gux-table-toolbar-action/gux-table-toolbar-action.tsx:
```diff
@@ -0,0 +1,90 @@
+  icon: string;
+
+  @State()
+  label: string;
```
is this used?

daragh-king-genesys commented on src/components/beta/gux-table-toolbar/gux-table-toolbar-action/gux-table-toolbar-action.tsx:
```diff
@@ -0,0 +1,90 @@
+
+  private returnActionLocale(): string {
+    return this.i18n(
+      `action${this.action.charAt(0).toUpperCase() + this.action.slice(1)}`
```
I would create a capitalise first letter util function and import it.

daragh-king-genesys commented on src/components/beta/gux-table-toolbar/gux-table-toolbar-action/gux-table-toolbar-action.tsx:
```diff
@@ -0,0 +1,90 @@
+    return (
+      <gux-button accent={this.primary ? 'primary' : 'secondary'} type="button">
+        <gux-icon
+          iconName={this.action == 'revert' ? 'reset' : this.action}
```
I would not have this much logic in the template. I would create and call a function instead

daragh-king-genesys commented on src/components/beta/gux-table-toolbar/gux-table-toolbar-action/gux-table-toolbar-action.tsx:
```diff
@@ -0,0 +1,90 @@
+          decorative
+        />
+        <span class={{ 'gux-sr-only': this.iconOnly }}>
+          {this.returnActionLocale()}
```
I would have expected this to be 
```
<slot>{this.returnActionLocale()}</slot>
```
so devs can give an more specific name if the preset buttons if they wanted

daragh-king-genesys commented on src/components/beta/gux-table-toolbar/gux-table-toolbar-menu-button/gux-table-toolbar-menu-button-popover/gux-table-toolbar-menu-button-popover.tsx:
```diff
@@ -0,0 +1,68 @@
+      </div>
+    ) as JSX.Element;
+  }
+}
```
Why not use `gux-popup`?

gavin-everett-genesys replied:
I think back then when I was implementing it I was having css issues and could not figure out why the stylings were not working when using it so I created my own version. Turns out the reason it didn't work in the first place was because there was a missing `.less` on my styleUrl. It works fine now 🙈

gavin-everett-genesys commented:
Addressed PR comments above,

The toolbar operates pretty well now when resizing. There is only one edge case that I could not solve but I dont think it will be a PR blocker. I show the issue at the end of the video if you are in the `iconOnly` layout and refresh sometimes it will automatically shift to the `condensed` layout. I am still looking into this issue but hopefully it isnt a blocker.

https://user-images.githubusercontent.com/101644078/189103736-00b71148-93f8-45eb-8288-db82544fc27e.mov

MattCheely reviewed (changes requested):
I think the menu collapse logic won't work as it exists.

MattCheely commented on src/components/beta/gux-table-toolbar/example.html:
```diff
@@ -0,0 +1,51 @@
+    <gux-table-toolbar-menu-button>
+      <gux-list-item onclick="notify(event)">Modify</gux-list-item>
+      <gux-list-item onclick="notify(event)">Extract</gux-list-item>
+    </gux-table-toolbar-menu-button>
```
To elaborate on this a little bit more, The way I'm thinking it could work is that the `menuActions` slot is always rendered in a dropdown, but the locations where the `contextual` and `permanent` slots go depends on the `displayedLayout` property, such that we move them into the dropdown as siblings of the `menuActions` slot in the condensed mode. From there it's a fair bit of JS and CSS to make everything look & work right, but I think it's the most reasonable options.

gavin-everett-genesys commented:
Following my recent push,

I have now removed the condensed menu layout and have created a ticket :https://inindca.atlassian.net/browse/COMUI-1202 to look into this next week. I also had a discussion with @daragh-king-genesys  about the API and agreed that we should seperate preset actions from custom actions. The are now two options for actions,

1.
`<gux-table-toolbar-action action="export"></gux-table-toolbar-action>`
You can use this and provide a prop called action and choose one of the presets of your choice.
2.
`  <gux-table-toolbar-custom-action>
      <span slot="action">Add</span>
      <gux-icon slot="icon" icon-name="add" screenreader-text="add"></gux-icon>
    </gux-table-toolbar-custom-action> `
If you want to have a custom action then you would use the above which takes a slotted icon and span.

I think this is a more cleaner API then what was previously implemented.

I had to use named slots as if both the label and icon are in the same slot I dont think there is a way to hide the label when you are resizing.

daragh-king-genesys commented on src/components/beta/gux-table-toolbar/gux-table-toolbar.tsx:
```diff
@@ -0,0 +1,167 @@
+        this.displayedLayout = 'iconOnly';
+        this.collapseActions();
+      }
+    });
```
Can you add a few inline comments to this function explaining what is happening.

For example I am confused why you check `this.displayedLayout === 'full'` before updating `expandAt`.
I am also wondering why you need to store `expandAt` at all.

MattCheely replied:
Good point, and this is on me, as I came up with this logic.  `expandAt` tracks the size at which the `full` layout should be displayed (or should be tested again).  There's no way to know that size in advance, since it depends on the number of buttons in the component. So the component starts off trying the `full` display and shrinks if there's not enough space for it. The last seen size where `full` was displayed is tracked in `expandAt`, and when the component is in icon-only mode, it will try to expand when it reaches that size again. That will either work (in which case all is good) or fail, increasing the size of `expandAt`.

daragh-king-genesys commented on src/components/beta/gux-table-toolbar/gux-table-toolbar.tsx:
```diff
@@ -0,0 +1,167 @@
+
+  render(): JSX.Element {
+    return (
+      <Host role="toolbar" aria-orientation="horizontal">
```
I would not use the `Host` element unless you have to and I don't think you do need to do it here.

gavin-everett-genesys commented on src/components/beta/gux-table-toolbar/gux-table-toolbar-action/gux-table-toolbar-action.tsx:
```diff
@@ -0,0 +1,52 @@
+
+  @Prop()
+  accent: 'primary' | 'secondary' = 'secondary';
+
```
I dont think this is needed as the resize is based off the `gux-table-toolbar-custom-action`

gavin-everett-genesys commented on src/components/beta/gux-table-toolbar/gux-table-toolbar-custom-action/gux-table-toolbar-custom-action.tsx:
```diff
@@ -0,0 +1,70 @@
+    return (
+      <gux-button accent={this.accent} type="button">
+        {this.renderActionIcon()}
+        {!this.hasActionSlot ? this.renderActionTitle() : null}
```
I agree initially what was happening was that if you were just to use an icon by itself the `span` for the `text slot` would still render causing the icon to be shifted to the left by 4px. I thought this might be a good way but found a better way through css. I just check if the slot is empty through `:empty` and if so I remove the padding.

gavin-everett-genesys commented:
I have now implemented the new solution for the condensed layout and I think its covers most edge cases. Props to @MattCheely for the idea of moving the location of the slots depending on the layout it works pretty well.

In this update,

**1.**
Instead of passing `gux-list-item` into the `menuActions` slot I am passing `gux-table-toolbar-action` instead. I currently have the ghost accent applied when they are moved into the condensed menu layout which matches the styling set out in the design guide the only difference is that the current ghost button is only styled in blue. So I think that a black accent ghost button would be perfect here and would rule out any css needed to be applied,
<img width="408" alt="ghost_button_black" src="https://user-images.githubusercontent.com/101644078/192565501-d5d79146-f9da-4a8a-8fa3-e5f80ad3e420.png">

Accessibility also will need to be done on the list.

**2.**
The resizing is based off the viewport as was specified in the design docs this can change though if needed. There was quite a bit of tweaking to get it right with the movement of the slots.

**3.**
Centralised some functions.

daragh-king-genesys commented on src/components/beta/gux-table-toolbar/gux-table-toolbar.service.tsx:
```diff
@@ -0,0 +1,81 @@
+    | HTMLGuxTableToolbarCustomActionElement[]
+    | HTMLGuxTableToolbarCustomActionElement,
+  accent: GuxTableToolbarActionAccent
+): GuxTableToolbarActionAccent {
```
This should be returning `void`. The others as well I think.

daragh-king-genesys replied:
Bumping this comment

daragh-king-genesys commented on src/components/beta/gux-table-toolbar/gux-table-toolbar.tsx:
```diff
@@ -0,0 +1,213 @@
+ * @slot searchFilter - Slot for search and filter.
+ * @slot contextual - Slot for contextual actions.
+ * @slot permanent - Slot for permanent actions.
+ * @slot primary - Slot for a primary action.
```
missing menuActions slot

daragh-king-genesys commented on src/components/beta/gux-table-toolbar/gux-table-toolbar.tsx:
```diff
@@ -0,0 +1,213 @@
+
+  get contextualSlot(): HTMLElement | null {
+    return this.root.querySelector('[slot="contextual"]');
+  }
```
Would a 'getSlot(slotName)' function be simpler?

daragh-king-genesys commented on src/components/beta/gux-table-toolbar/gux-table-toolbar.tsx:
```diff
@@ -0,0 +1,213 @@
+    return Array.from(
+      this.permanentSlot?.querySelectorAll('gux-table-toolbar-custom-action')
+    );
+  }
```
See will the assignedNode work for this.
`getSlot(slotName).assignedNodes()`

daragh-king-genesys commented on src/components/beta/gux-table-toolbar/gux-table-toolbar.tsx:
```diff
@@ -0,0 +1,213 @@
+  }
+
+  @OnResize()
+  checkResponsiveLayout(): void {
```
This function is very large and complex. See can you breakout parts out into named functions to improve readability.

MattCheely commented:
This is looking promising!

gavin-everett-genesys commented on src/components/beta/gux-table-toolbar/gux-table-toolbar.tsx:
```diff
@@ -0,0 +1,213 @@
+        )
+      );
+    }
+  }
```
We could probably doing something similar like what was done in `form-field-number` using the ::part element.

daragh-king-genesys commented on src/components/beta/gux-table-toolbar/gux-table-toolbar-menu-button/gux-table-toolbar-menu-button.tsx:
```diff
@@ -0,0 +1,132 @@
+  render(): JSX.Element {
+    return (
+      <div class="gux-table-toolbar-menu-container">
+        <gux-popup expanded={this.isOpen}>
```
for consistency I would rename `isOpen` to `expanded`

daragh-king-genesys commented on src/components/beta/gux-table-toolbar/gux-table-toolbar-menu-button/gux-table-toolbar-menu-button.tsx:
```diff
@@ -0,0 +1,132 @@
+
+  render(): JSX.Element {
+    return (
+      <div class="gux-table-toolbar-menu-container">
```
Is this wrapper needed?

daragh-king-genesys commented on src/components/beta/gux-table-toolbar/gux-table-toolbar-menu-button/gux-table-toolbar-menu-button.tsx:
```diff
@@ -0,0 +1,132 @@
+@Component({
+  styleUrl: 'gux-table-toolbar-menu-button.less',
+  tag: 'gux-table-toolbar-menu-button',
+  shadow: true
```
Should this delegate focus

daragh-king-genesys commented on src/components/beta/gux-table-toolbar/gux-table-toolbar-menu-button/gux-table-toolbar-menu-button.tsx:
```diff
@@ -0,0 +1,132 @@
+              <button
+                type="button"
+                ref={el => (this.dropdownButton = el)}
+                onMouseUp={() => this.toggle()}
```
Why onMouseUp?

daragh-king-genesys commented on src/components/beta/gux-table-toolbar/gux-table-toolbar.service.tsx:
```diff
@@ -0,0 +1,81 @@
+  actions:
+    | HTMLGuxTableToolbarCustomActionElement[]
+    | HTMLGuxTableToolbarCustomActionElement
+): boolean {
```
Why are these returning a boolean?

daragh-king-genesys commented on src/components/beta/gux-table-toolbar/gux-table-toolbar.tsx:
```diff
@@ -0,0 +1,251 @@
+export class GuxTableToolbar {
+  /**
+   * Reference to the host element
+   */
```
Pointless comment

daragh-king-genesys commented on src/components/beta/gux-table-toolbar/gux-table-toolbar.tsx:
```diff
@@ -0,0 +1,251 @@
+    return this.filterActions.concat(this.contextualActions);
+  }
+
+  private styleSearchField(): void {
```
can we not do this with slotted styles?

gavin-everett-genesys replied:
I dont think this is possible as I would need to style the `input` inside the shadowDOM of form-field-search. The only way I could think of is ::part.

daragh-king-genesys commented on src/components/beta/gux-table-toolbar/gux-table-toolbar.service.tsx:
```diff
@@ -0,0 +1,81 @@
+import { isArray } from 'vega';
+import { GuxTableToolbarActionAccent } from './gux-table-toolbar-action-accents.types';
+
+export function setAccent(
```
You don't have to make this change but I would probably try to merge all these functions into one or two. Something like
```
export function bulkSetIconOnly(
  targets: (HTMLGuxTableToolbarCustomActionElement[] | HTMLGuxTableToolbarCustomActionElement)[]
    | HTMLGuxTableToolbarCustomActionElement[]
    | HTMLGuxTableToolbarCustomActionElement,
  value: boolean
): void {
  [].concat(targets).flat().forEach(target => {
    target['iconOnly'] = value;
  })
}
```

daragh-king-genesys commented on src/components/beta/gux-table-toolbar/gux-table-toolbar.tsx:
```diff
@@ -0,0 +1,230 @@
+
+      if (
+        elRect.right >
+        (window.innerWidth || document.documentElement.clientWidth)
```
@gavin-everett-genesys is this comment addressed yet?

daragh-king-genesys reviewed (dismissed):
Looking very good now. I have a few trivial comments and only Matts comment on resizing seems to be left.

daragh-king-genesys commented:
@gavin-everett-genesys is this good yo merge now?
